### PR TITLE
Add GetAllKeysByPrefix to InMemory and Redis

### PR DESCRIPTION
**Pull request by meysamhadeli**
Hi.
I added this feature #397 to EasyCahing.

- Add getall (inMemory - redis)
- Add getall-keys (inMemory - redis)

### CODE REVIEW

**Comment by meysamhadeli** [comment]
Hi.
Is this any problem with this pull request?

**Comment by catcherwong** [comment]
Still in the review, maybe need one more day.

**Comment by meysamhadeli** [comment]
> Still in the review, maybe need one more day.

Ok, Thanks :)

File: src/EasyCaching.LiteDB/DefaultLiteDBCachingProvider.cs [hunk 1]
```diff
@@ -321,6 +321,16 @@ public override void BaseSetAll<T>(IDictionary<string, T> values, TimeSpan expir
             return GetDict<T>(list);
         }
 
+        public override IDictionary<string, CacheValue<T>> BaseGetAll<T>()
+        {
+            throw new NotImplementedException();
```
Comment (catcherwong) [inline src/EasyCaching.LiteDB/DefaultLiteDBCachingProvider.cs:326]: Is `NotSupportedException` better than `NotImplementedException` or not?

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs [hunk 2]
```diff
@@ -295,6 +295,56 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
+        /// Gets all async.
+        /// </summary>
+        /// <returns>The all async.</returns>
+        /// <param name="cancellationToken">CancellationToken</param>
+        /// <typeparam name="T">The 1st type parameter.</typeparam>
+        public override async Task<IDictionary<string, CacheValue<T>>> BaseGetAllAsync<T>(CancellationToken cancellationToken = default)
+        {
+            var connection = _cache.Multiplexer;
+
+            var keyArray = connection?.GetServer(connection?.GetEndPoints().FirstOrDefault())
```
Comment (catcherwong) [inline src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:309]: Here should use `_servers`.

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs [hunk 3]
```diff
@@ -295,6 +295,56 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
+                ?.Keys(database: _cache.Database, pattern: "*")?.ToArray();
+
+            if (keyArray?.Length == 0 || keyArray is null)
+                return new Dictionary<string, CacheValue<T>>();
+
+            var values = _cache.StringGet(keyArray?.Select(k => (RedisKey)k).ToArray());
```
Comment (catcherwong) [inline src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:315]: Here should consider the key prefix.

**Review by catcherwong — CHANGES_REQUESTED** [review]

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs [hunk 4]
```diff
@@ -295,6 +295,56 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
+        /// Gets all keys async.
+        /// </summary>
+        /// <returns>The all keys async.</returns>
+        public override async Task<IEnumerable<string>> BaseGetAllKeysAsync(CancellationToken cancellationToken = default)
+        {
+            var keyArray = new List<string>();
+
+            var connection = _cache.Multiplexer;
+
+            var redisKeys = connection?.GetServer(connection.GetEndPoints()?.FirstOrDefault())
```
[the remark this answers is not in the document]
Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:340]: Ok. I fix that.

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs (hunk 2 above)
Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:309]: Ok. I fix that.

File: src/EasyCaching.LiteDB/DefaultLiteDBCachingProvider.cs (hunk 1 above)
Comment (meysamhadeli) [reply src/EasyCaching.LiteDB/DefaultLiteDBCachingProvider.cs:326]: Yes. It's better. I change it.

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs [hunk 5]
```diff
@@ -295,6 +295,56 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
+                ?.Keys(database: _cache.Database, pattern: "*")?.ToArray();
+            
+            keyArray.AddRange(redisKeys?.Select(key => (string) key) ?? new List<string>());
```
[the remark this answers is not in the document]
Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs]: I refactor this method with optional prefix.

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs (hunk 3 above)
Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs]: I refactor this method with optional prefix.

**Comment by meysamhadeli** [comment]
I refactor all issue in my code and sent a pull request for that :)

Full changeset is here:
https://github.com/dotnetcore/EasyCaching/pull/422/commits/8c70b43646b53a191d1453549cf37af0d081eba0

**Comment by meysamhadeli** [comment]
Did you have time to check the latest changes?

File: src/EasyCaching.InMemory/DefaultInMemoryCachingProvider.Async.cs [hunk 6]
```diff
@@ -275,6 +275,32 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
             return await Task.FromResult(_cache.GetAll<T>(cacheKeys));
         }
+        
+        /// <summary>
+        /// Gets all async.
+        /// </summary>
+        /// <returns>The all async.</returns>
+        /// <param name="cancellationToken">CancellationToken</param>
+        /// <typeparam name="T">The 1st type parameter.</typeparam>
+        public override async Task<IDictionary<string, CacheValue<T>>> BaseGetAllAsync<T>(string prefix = "", CancellationToken cancellationToken = default)
```
Comment (catcherwong) [inline src/EasyCaching.InMemory/DefaultInMemoryCachingProvider.Async.cs:285]: Parameter 'prefix' has no matching param tag in the XML comment

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (catcherwong) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs]: What I mean is not optional prefix. 

We should consider the key prefix that user set it global in their code.

https://github.com/StackExchange/StackExchange.Redis/blob/02c29ef8e8ddf8de9556c113a8a6fb1563f46a83/src/StackExchange.Redis/KeyspaceIsolation/DatabaseExtension.cs#L40

https://github.com/dotnetcore/EasyCaching/blob/5b093dc309989dbafb92d050be65ccddd3876c6a/src/EasyCaching.Redis/DefaultRedisCachingProvider.cs#L329-L332

[the remark this answers is not in the document]
Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs]: Ok, I remove all method with optional prefix as you mentioned and only add these two methods:
- GetAllKeysByPrefix
- GetAllKeysByPrefixAsync

**Comment by meysamhadeli** [comment]
I removed methods with `optional prefix parameter`  and only add these two methods:

- GetAllKeysByPrefix
- GetAllKeysByPrefixAsync

Full changeset is here:
https://github.com/dotnetcore/EasyCaching/pull/422/commits/d40f6bf7a71dbe09e12009a753813c6e73a44bc5

File: src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs [hunk 7]
```diff
@@ -295,6 +295,27 @@ public override async Task BaseSetAllAsync<T>(IDictionary<string, T> values, Tim
...
+        /// <returns>The all keys by prefix async.</returns>
+        public override async Task<IEnumerable<string>> BaseGetAllKeysByPrefixAsync(string prefix, CancellationToken cancellationToken = default)
+        {
+            ArgumentCheck.NotNullOrWhiteSpace(prefix, nameof(prefix));
+
+            prefix = this.HandlePrefix(prefix);
+
+            var redisKeys = this.SearchRedisKeys(prefix);
+            
+            var result = redisKeys?.Select(key => (string) key)?.Distinct();
```
Comment (catcherwong) [inline src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:314]: Suppose we do not set a global prefix.

We set something as below

```
Set("order:aaaa", "1", TimeSpan.FromDay(1));
Set("order:aaab", "1", TimeSpan.FromDay(1));
Set("order:aaac", "1", TimeSpan.FromDay(1));
```

When we call `GetAllKeysByPrefixAsync("order", default)`,  the result will be  `["order:aaaa", "order:aaab", "order:aaac"]`

This is okey.

But, suppose we do set a global prefix.

```
services.AddEasyCaching(x =>
 {            
                x.UseRedis(options =>
                {
                    options.DBConfig = new RedisDBOptions
                    {
                        KeyPrefix = "order:"
                    };
                   // ....
                }, "WithKeyPrefix");
            });
```

Then, we set something as below

```
Set("order:aaaa", "1", TimeSpan.FromDay(1));
Set("order:aaab", "1", TimeSpan.FromDay(1));
Set("order:aaac", "1", TimeSpan.FromDay(1));
```

When we call `GetAllKeysByPrefixAsync("order", default)`,  which result do you expect?

 `["order:aaaa", "order:aaab", "order:aaac"]` or  `["order:order:aaaa", "order:order:aaab", "order:order:aaac"]`

Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:314]: Hi,
I Think right result is best: `["order:order:aaaa", "order:order:aaab", "order:order:aaac"]`. we add `"order"` from `GetAllKeysByPrefixAsync("order", default)` to base prefix that here is `order:` and result `order:  + order` and we have `order:order` in start key prefix. We consider `base prefix first` and combine it with `parameter prefix` from method. If you are agree with this approach I change my code to this solution.

Comment (catcherwong) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:314]: When you get this result `["order:order:aaaa", "order:order:aaab", "order:order:aaac"]`, and if you want to get one of the value of them, you will call `Get<T>("order:order:aaaa")`, but it cannot get the value if you set a base/global  preifx, at this time `["order:aaaa", "order:aaab", "order:aaac"]` is a better result.

If you only want to get the cache key store in redis,  `["order:order:aaaa", "order:order:aaab", "order:order:aaac"]` is ok.

If you want to get the cache key and call other methods provided by EasyCaching, such as `Get<T>`,  `["order:aaaa", "order:aaab", "order:aaac"]` is ok.

This is why I suggest you pay attention to deal with the key prefix.

Comment (meysamhadeli) [reply src/EasyCaching.Redis/DefaultRedisCachingProvider.Async.cs:314]: I wanna get all cache keys and using this keys in other method EasyCaching like `Get<T>` and for handle this scenario I change my code and now if we call `GetAllKeysByPrefixAsync("order", default)` and have `base key prefix = order:`, we  get this result `["order:aaaa", "order:aaab", "order:aaac"]` as well with calling this method `Get<T>`.

Full changeset is here:
cd5d01f1a1f4e31649507380aec605742fe55b29

**Pull request merged**